benchrun/python/benchrun/_benchmark.py

```diff
@@ -0,0 +1,263 @@
+        queue : multiprocessing.Queue
+            An instance of `multiprocessing.Queue` used for sending data back to a parent process.
+        """
+
+        if settings["drop_caches"]:
+            self.cache.sync_and_drop()
+        if settings["gc_collect"]:
+            gc.collect()
+        if settings["gc_disable"]:
+            gc.disable()
+
+        elapsed_time = None
+        error = None
+        try:
+            setup_results = self.setup(case=case)
+            start_time = time.time()
+            run_results = self.run(case=case, setup_results=setup_results)
+            end_time = time.time()
+            self.teardown(case=case, run_results=run_results)
+            elapsed_time = end_time - start_time
+        except Exception as e:
+            error = repr(e)
+            warnings.warn(error)
+
+        gc.enable()
+
+        queue.put({"time": elapsed_time, "error": error})
+        return elapsed_time
+
+
+class Benchmark:
+    """
+    A class that takes an `Iteration` instance and runs it correctly for each case
+
+    Parameters
+    ----------
+    iteration : Iteration
+        An instance of `Iteration` defining code to benchmark
+    case_list : CaseList
+        An instance of `CaseList` defining a list of valid cases for which to run this benchmark.
+    results_fields_append : Dict[str, Any]
+        A dict of values to be appended to `BenchmarkResult` values after
+        instantiation. Useful for appending extra tags or other metadata in addition
+        to that gathered elsewhere. Only applicable for dict attributes. For each
+        element, will override any keys that already exist, i.e. it does not append
+        recursively.
+
+        In most cases, should be set on the adapter, not here, but can be set here to
+        specify benchmark-dependent values like appending a tag for ``suite``.
+    drop_caches : bool
+        Try to drop disk caches?
+    gc_collect : bool
+        Run garbage collection before timing code?
+    gc_disable : bool
+        Disable garbage collection during timing?
+    error_handling : str
+        What should happen if a benchmark errors out? Options: ``"stop"`` (skip future iterations
+        and report only error), ``"break"`` (skip future iterations and report everything so far),
+        ``"continue"`` (run all iterations even if they fail and report everything). As we currently
+        can't report both metrics and errors for the same benchmark, ``"stop"`` and ``"break"`` are
+        currently identical, and ``"continue"`` may run longer, but will report the same thing.
+    """
+
+    iteration: Iteration = None
+    case_list: CaseList = None
+    result_fields_append: Dict[str, Any] = None
+    settings = None
+    cache = None
+
+    def __init__(
+        self,
+        iteration: Iteration,
+        case_list: CaseList,
+        result_fields_append: Dict[str, Any] = {},
+        drop_caches: bool = False,
```

Comment:
I think we want this to be true? But really we should _also_ have a TODO that does a really investigation into how this impacts the actual data.

Reply:
[The current default is false](https://github.com/conbench/conbench/blob/main/conbench/runner.py#L343), which is why I set it this way. We can change that, but it may cause a discontinuity in some benchmarks that weren't explicitly setting it (whenever we get around to porting them).

Reply:
Let's keep this as whatever the current default is and mark a TODO to come back and actually figure out what the right answer is. I think we've cargo-culted ourselves on this one, but that's an empirical question we should dedicate the time to figuring out in the future.